cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/DistributedStreamService.java

```diff
@@ -381,11 +377,25 @@ private void performLeaderElection() {
       zkClient, "/election/" + STREAMS_COORDINATOR, new ElectionHandler() {
       @Override
       public void leader() {
-        LOG.info("Became Stream handler leader. Starting resource coordinator.");
-        resourceCoordinator = new ResourceCoordinator(getCoordinatorZKClient(), discoveryServiceClient,
-                                                      new BalancedAssignmentStrategy());
-        resourceCoordinator.startAndWait();
-        updateRequirement();
+        while (true) {
+          LOG.info("Became Stream handler leader. Starting resource coordinator.");
+          heartbeatsSubscriptionExecutor = Executors.newSingleThreadExecutor(
+            Threads.createDaemonThreadFactory("heartbeats-subscription-executor"));
+          try {
+            heartbeatsSubscription = subscribeToHeartbeatsFeed();
```

Comment:
Ok so I think the right way here is to only put into a loop ``subscribeToHeartbeatsFeed``. But then, if an exception occurs in another method, say ``updateRequirement``, then what should be the expected behavior? If the exception if propagated, then will the stream service will just fail? Or will zookeeper retry the leader election? How does it work?